### PR TITLE
OneStage

### PR DESCRIPTION
**Pull request by Jinp800125**
OneStage
  - Gemm
  - GroupGemm
    - w/ w/o UserArg
  - Test case

### CODE REVIEW

File: tensilelite/Tensile/AsmStoreState.py [hunk 1]
```diff
@@ -196,16 +200,21 @@ def __init__(self, kernelWriter, kernel, gwvw, edge, beta, atomic, elements):
...
             self.numAddrVgpr = 1
             self.sharedColDVgprs = kernelWriter.vgprPool.checkOut(1, "sharedColDVgprs")
+            if kernel["_GlobalAccumulation"] == "MultipleBufferSingleKernel":
+                self.sharedGSUSyncVgprs = kernelWriter.vgprPool.checkOut(1, "sharedGSUSyncVgprs")
+            else:
+                self.sharedGSUSyncVgprs = None
             self.singleColBiasAddrUpdated = False
             self.singleColEAddrUpdated    = False
             self.singleColDAddrUpdated    = False
+            self.singleColTDAddrUpdated    = False
```
Comment (hcman2) [inline tensilelite/Tensile/AsmStoreState.py]: one extra space here

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 2]
```diff
@@ -201,7 +501,7 @@ def _prolog(self, module: Module):
                                for idx, element in enumerate(self.batchElements)])
     module.addComment2(commentStr)
 
-    self.ss.setupStoreElementsForBatch(self.kernel, self.gwvw, self.batchElements, self.batchElementSgprs, isOptNLL=False, biasDim=self.biasDim)
+    self.ss.setupStoreElementsForBatch(self.kernel, self.gwvw, self.batchElements, self.batchElementSgprs, isOptNLL=True, biasDim=self.biasDim)
```
Comment (hcman2) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py]: Why set True here?

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 3]
```diff
@@ -436,10 +801,13 @@ def _epilog(self, module: Module):
     checkedDataBias = {}
     checkedDataScaleAlphaVec = {}
     for elementIdx in range(len(self.batchElements)):
+      sumIdxGSUSYNC = self.ss.elementSumIdx[elementIdx]
+      addrCalc: AddrCalculation = self.ss.elementAddr[elementIdx]
```
Comment (hcman2) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py]: line 805 can be removed?

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 4]
```diff
@@ -976,25 +1351,55 @@ def _emitNonatomicAdd(self, module: Module):
...
-        tmpStoreCode = self.parentWriter.addStore(self.kernel, self.ss, 'D', addrCalc, sumIdx, self.tmpS01, self.edge, comment="store D")
+        if self.kernel["_GlobalAccumulation"] == "MultipleBufferSingleKernel":#GSUGSU
+          tmpStoreCode = self.parentWriter.addStore(self.kernel, self.ss, 'TD', addrCalc, sumIdx, self.tmpS01, self.edge, comment="store TD not StoreRemapVectorWidth")
+        else:
+          tmpStoreCode = self.parentWriter.addStore(self.kernel, self.ss, 'D', addrCalc, sumIdx, self.tmpS01, self.edge, comment="store D")
         if self.kernel["GroupLoadStore"]:
           storeCode.add(tmpStoreCode)
         else:
           module.add(tmpStoreCode)
+          # module.add(SNop(0, "1 wait state required when next inst writes vgprs held by previous dwordx4 store inst"))
```
Comment (hcman2) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py]: This is commented. Can it be removed?

File: tensilelite/Tensile/Components/GlobalWriteBatch.py (hunk 2 above)
Comment (hcman2) [reply tensilelite/Tensile/Components/GlobalWriteBatch.py]: We can remove this parameter "isOptNLL".

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 5]
```diff
@@ -1446,7 +1491,10 @@ def calculateWG():
           module.addComment1("Check if custom structure pointer is null")
           module.add(SCmpEQU32(src0=sgpr("ArgType"), src1=2, comment="ArgType == 2 ?"))
           module.add(SCBranchSCC1(labelName=extValidLabel.getLabelName(), comment="branch if ArgType == 2"))
-          module.add(SMovB32(dst=sgpr(tmpSgprArgOffsett), src=(self.argLoader.getOffset() + (numStoreSgprToLoad * 4))))
+          if ((kernel["GlobalSplitUAlgorithm"] == 'MultipleBufferSingleKernel')):
+            module.add(SMovB32(dst=sgpr(tmpSgprArgOffsett), src=(self.argLoader.getOffset() + (numStoreSgprToLoad * 4) + (5)*4)))
```
Comment (hcman2) [inline tensilelite/Tensile/KernelWriterAssembly.py]: What is this magic number 5? Could you use a variable or add a comment?

File: tensilelite/Tensile/SolutionStructs.py [hunk 6]
```diff
@@ -3575,6 +3579,20 @@ def subCheckLdsBlockSizePerPad(tc, idx):
...
+        reject(state, "Occupancy limit!! overflowed resources MultipleBufferSingleKernel does not support")
+      if state["ProblemType"]["UseScaleCD"]:
+        reject(state, "MultipleBufferSingleKernel not support UseScaleCD yet")
+      if state["ProblemType"]["UseE"]:
+        reject(state, "MultipleBufferSingleKernel not support UseE yet")
+      if state["ProblemType"]["BiasSrc"] != "D":
+        reject(state, "MultipleBufferSingleKernel not support BiasSrc not D yet")
+      if state["ProblemType"]["DataType"].isDouble():
+        reject(state, "MultipleBufferSingleKernel not support " + str(state["ProblemType"]["DataType"])  + " yet")
+      if state["Sparse"] == 2:
```
Comment (hcman2) [inline tensilelite/Tensile/SolutionStructs.py]: state["ProblemType"]["Sparse"]

Comment (Jinp800125) [reply tensilelite/Tensile/SolutionStructs.py]: fixed

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 5 above)
Comment (Jinp800125) [reply tensilelite/Tensile/KernelWriterAssembly.py]: modified

File: tensilelite/Tensile/Components/GlobalWriteBatch.py (hunk 4 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Components/GlobalWriteBatch.py]: removed

File: tensilelite/Tensile/Components/GlobalWriteBatch.py (hunk 3 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Components/GlobalWriteBatch.py]: removed

File: tensilelite/Tensile/AsmStoreState.py (hunk 1 above)
Comment (Jinp800125) [reply tensilelite/Tensile/AsmStoreState.py]: removed

File: tensilelite/Tensile/Components/GlobalWriteBatch.py (hunk 2 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Components/GlobalWriteBatch.py]: separated

File: tensilelite/Tensile/Source/client/main.cpp [hunk 7]
```diff
@@ -136,6 +136,7 @@ namespace Tensile
                 ("init-scaleC",              po::value<InitMode>()->default_value(InitMode::Two), "Initialization for scaleC")
                 ("init-scaleD",              po::value<InitMode>()->default_value(InitMode::Two), "Initialization for scaleD")
                 ("init-scaleAlphaVec",           po::value<InitMode>()->default_value(InitMode::One), "Initialization for scaleAlphaVec")
+                ("init-Synchronizer",           po::value<InitMode>()->default_value(InitMode::Zero), "Initialization for Synchronizer")
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/client/main.cpp]: align previous line

File: tensilelite/Tensile/KernelWriter.py [hunk 8]
```diff
@@ -3649,6 +3659,12 @@ def readWriteVectors(mat, vw, kernel):
 
     self.defineSgpr("GSU", 1)  # Can't move to the front because of the preload arguments
 
+    if kernel["ProblemType"]["GroupedGemm"]:
+      if GSUAMBSK:
+        self.defineSgpr("GSUSync", 1)
+        self.defineSgpr("AddressTC", numSgprAddressD)
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriter.py]: TC or TD??

Comment (Jinp800125) [reply tensilelite/Tensile/KernelWriter.py]: modified

File: tensilelite/Tensile/Source/client/main.cpp (hunk 7 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Source/client/main.cpp]: modified

File: tensilelite/Tensile/AsmAddressCalculation.py [hunk 9]
```diff
@@ -634,24 +646,44 @@ def incrementToNextRow(self, kernel, tc, ss, stmp, bpeType=None):
...
-                    module.add(SSubU32(dst=sgpr("Srd%s+0"%(tc)), \
-                                        src0=sgpr("Srd%s+0"%(tc)), \
-                                        src1=sgpr(stmp), \
-                                        comment="incToNextRow: gra SRD -= inc(lower)" ))
-                    module.add(SSubBU32(dst=sgpr("Srd%s+1"%(tc)), \
-                                        src0=sgpr("Srd%s+1"%(tc)), \
-                                        src1=0, \
-                                        comment="incToNextRow: gra SRD -= inc(upper)" ))
+                if dst!= -1:
+                    if numRows >= 0:
```
Comment (jichangjichang) [inline tensilelite/Tensile/AsmAddressCalculation.py]: Better not to produce redundant codes
if dst == -1:
   dst== "Srd%s"%(tc)
...
if numRows>=0

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 10]
```diff
@@ -7416,7 +7457,7 @@ def computeStoreSrdStart(self, kernel, srdTcList: list, sgprBpeList = [], useSiz
           tmpSgpr1 = tmpSgprInfo.idx+2
           tmpSgpr2 = tmpSgprInfo.idx+0
           tmpSgpr3 = tmpSgprInfo.idx+3
-        module.addComment("GSU Output Buffer offset: Free0 + (Free1-1)*StrideC1J + (Free2-1)*StrideCK * GSUIdx * bpe%s")
+        module.addComment("GSU Output Buffer offset: Free0 + (Free1-1)*StrideC1J + (Free2-1)*StrideCK * GSUIdx * bpe%s" % (self.sgprPool.size()))
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: bpe<SGPR pool size>?

File: tensilelite/Tensile/Common.py [hunk 11]
```diff
@@ -514,7 +514,7 @@ def getArchitectureName(gfxName):
     # choose how to do GlobalSplitU
     # 1: use atomic operation to accumulate on one buffer
```
Comment (jichangjichang) [inline tensilelite/Tensile/Common.py:515]: add description

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 12]
```diff
@@ -181,6 +182,297 @@ def _chooseAddForAtomic(self, kernel, dst, src0, src1, comment):
...
+    module.add(SSubU32(dst=sgpr(tmpS06+0), \
+                                    src0=sgpr("SrdD+0"), \
+                                    src1=sgpr(tmpS05+0), \
+                                    comment="" ))
+    module.add(SSubBU32(dst=sgpr(tmpS06+1), \
+                        src0=sgpr("SrdD+1"), \
+                        src1=sgpr(tmpS05+1), \
+                        comment="" ))
+
+    module.add(SMovB32(sgpr(tmpS06+2), sgpr("SrdD+2"), "Move HELLO"))
```
Comment (jichangjichang) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py]: Hello, how are you?

Comment (Jinp800125) [reply tensilelite/Tensile/Components/GlobalWriteBatch.py]: Outdated, need refresh XD

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 10 above)
Comment (Jinp800125) [reply tensilelite/Tensile/KernelWriterAssembly.py]: Outdated, need refresh

File: tensilelite/Tensile/AsmAddressCalculation.py (hunk 9 above)
Comment (Jinp800125) [reply tensilelite/Tensile/AsmAddressCalculation.py]: done

File: tensilelite/Tensile/Common.py (hunk 11 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Common.py:515]: done

**Review by hcman2 — APPROVED** [review]

File: tensilelite/Tensile/Source/client/main.cpp [hunk 13]
```diff
@@ -135,7 +135,8 @@ namespace Tensile
                 ("init-scaleB",              po::value<InitMode>()->default_value(InitMode::Two), "Initialization for scaleB")
                 ("init-scaleC",              po::value<InitMode>()->default_value(InitMode::Two), "Initialization for scaleC")
                 ("init-scaleD",              po::value<InitMode>()->default_value(InitMode::Two), "Initialization for scaleD")
-                ("init-scaleAlphaVec",           po::value<InitMode>()->default_value(InitMode::One), "Initialization for scaleAlphaVec")
+                ("init-scaleAlphaVec",       po::value<InitMode>()->default_value(InitMode::One), "Initialization for scaleAlphaVec")
+                ("init-Synchronizer",        po::value<InitMode>()->default_value(InitMode::Zero), "Initialization for Synchronizer")
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/client/main.cpp]: Should not add this argument. synchronizer is solution dependent but not problem type dependent.

File: tensilelite/Tensile/Source/client/main.cpp [hunk 14]
```diff
@@ -252,6 +253,7 @@ namespace Tensile
                 ("use-scaleAB",               po::value<bool>()->default_value(false), "Use scaleAB.")
                 ("use-scaleCD",               po::value<bool>()->default_value(false), "Use scaleCD.")
                 ("use-scaleAlphaVec",         po::value<bool>()->default_value(false), "Use scaleAlphaVec.")
+                ("use-Synchronizer",                po::value<bool>()->default_value(false), "Use Synchronizer.")
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/client/main.cpp]: same as above descrption

Comment (Jinp800125) [reply tensilelite/Tensile/Source/client/main.cpp]: removed

File: tensilelite/Tensile/Source/client/main.cpp (hunk 13 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Source/client/main.cpp]: removed

Comment (Jinp800125) [reply tensilelite/Tensile/Source/client/main.cpp]: removed

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 15]
```diff
@@ -445,6 +445,25 @@ def defineVariableSgprs(self, kernel):
...
+    #########################################################
+    # Below calculates the number of sgprs needed not in epilogue
+    #########################################################
+    self.states.numStoreSgprNames2 = []
+    self.states.numStoreSgprNameSizes2 = []
+    storeSgprLoad2 = 0
+
+    if kernel["_GlobalAccumulation"] == 'MultipleBufferSingleKernel':
+      storeSgprLoad2 += self.states.rpga
+      # self.states.numStoreSgprNames2.append("AddressTC")
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: Remove unused codes

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 16]
```diff
@@ -7372,6 +7500,53 @@ def computeStoreVgprs(self, kernel):
...
   # globalWriteWorkGroupInit:
   ##############################################################################
+
+  def SrdTDinit(self, kernel):
+    module = Module("SrdTDinit")
+    tmpspgr0 = self.sgprPool.checkOut(1)
+    tmpspgr = self.sgprPool.checkOutAligned(2, 4, preventOverflow=False)
+
+    module.addSpaceLine()
+    # labelStr = self.labels.getNameInc("sgprSrdTC")
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: remove unused codes

Comment (Jinp800125) [reply tensilelite/Tensile/KernelWriterAssembly.py]: removed

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 15 above)
Comment (Jinp800125) [reply tensilelite/Tensile/KernelWriterAssembly.py]: removed

**Review by KKyang — APPROVED** [review]

**Review by jichangjichang — APPROVED** [review]
please merge after get 94x CI pass

**Pull request merged**